tsdb/index/tsi1/partition.go

```diff
@@ -897,6 +900,46 @@ func (p *Partition) compactionsEnabled() bool {
 	return p.compactionsDisabled == 0
 }
 
+func (p *Partition) runPeriodicCompaction() {
+	// kick off an initial compaction at startup without the optimization check
+	p.Compact()
+
+	// check for compactions once an hour (usually not necessary but a nice safety check)
+	t := time.NewTicker(p.MaxLogFileAge)
+	defer t.Stop()
+	for {
+		select {
+		case <-p.closing:
+			return
+		case <-t.C:
+			if p.NeedsCompaction() {
+				p.Compact()
+			}
+		}
+	}
+}
+
+// needsCompaction only requires a read lock and checks if there are files that could be compacted.
+// If compact is updated we should also update needsCompaction
+func (p *Partition) NeedsCompaction() bool {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if p.needsLogCompaction() {
+		return true
+	}
+	levelCount := make(map[int]int)
+	maxLevel := len(p.levels) - 2
+	// If we have 2 log files (level 0), or if we have 2 files at the same level, we should do a compaction.
+	for _, f := range p.fileSet.files {
+		level := f.Level()
+		levelCount[level] = levelCount[level] + 1
```

Comment:
Is this intended to be anything other than `levelCount[level] += 1` ?

Reply:
Made it `levelCount[level]++`